tools/pybind11Tools.cmake

```diff
@@ -18,24 +18,40 @@ find_package(PythonLibsNew ${PYBIND11_PYTHON_VERSION} REQUIRED)
 include(CheckCXXCompilerFlag)
 include(CMakeParseArguments)
 
+# Use the language standards abstraction if CMake supports it with the current compiler
+if(NOT CMAKE_VERSION VERSION_LESS 3.1)
+  if(NOT CMAKE_CXX_STANDARD)
+    if(CMAKE_CXX14_STANDARD_COMPILE_OPTION)
+      set(CMAKE_CXX_STANDARD 14)
+    elseif(CMAKE_CXX11_STANDARD_COMPILE_OPTION)
+      set(CMAKE_CXX_STANDARD 11)
+    endif()
+  endif()
+  if(CMAKE_CXX_STANDARD)
+    set(CMAKE_CXX_EXTENSIONS OFF)
+    set(CMAKE_CXX_STANDARD_REQUIRED ON)
+  endif()
+endif()
+
+# Fall back to heuristics 
 if(NOT PYBIND11_CPP_STANDARD AND NOT CMAKE_CXX_STANDARD)
-  if(NOT MSVC)
+  if(MSVC)
+    set(PYBIND11_CPP_STANDARD /std:c++14)
```

Comment:
MSVC 2015 Update 3 is the one that introduced `/std:c++14` and `/std:c++latest`.

https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler/

Reply:
Hm good point -- and I forgot that we're actually testing MSVC 2015 on AppVeyor.

Reply:
FWIW the MSVC behavior isn't changed in this PR from what it was before.  I only inverted the if/else order to have the larger, most often followed, code block to be the `if` condition rather than the `else`.